dap/adapter.go

```diff
@@ -0,0 +1,395 @@
+package dap
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+
+	"github.com/docker/buildx/build"
+	"github.com/google/go-dap"
+	gateway "github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+)
+
+type Adapter struct {
+	srv *Server
+	eg  *errgroup.Group
+	cfg build.InvokeConfig
+
+	initialized   chan struct{}
+	started       chan error
+	configuration chan struct{}
+
+	evaluateReqCh chan *evaluateRequest
+
+	threads      map[int]*thread
+	threadsMu    sync.RWMutex
+	nextThreadID int
+}
+
+func New() *Adapter {
+	d := &Adapter{
+		initialized:   make(chan struct{}),
+		started:       make(chan error, 1),
+		configuration: make(chan struct{}),
+		evaluateReqCh: make(chan *evaluateRequest),
+		threads:       make(map[int]*thread),
+		nextThreadID:  1,
+	}
+	d.srv = NewServer(d.dapHandler())
+	return d
+}
+
+func (d *Adapter) Start(ctx context.Context, conn Conn) error {
+	d.eg, _ = errgroup.WithContext(ctx)
+	d.eg.Go(func() error {
+		return d.srv.Serve(ctx, conn)
+	})
+
+	<-d.initialized
+	return <-d.started
+}
+
+func (d *Adapter) Stop() error {
+	if d.eg == nil {
+		return nil
+	}
+
+	d.srv.Go(func(c Context) {
+		c.C() <- &dap.TerminatedEvent{
+			Event: dap.Event{
+				Event: "terminated",
+			},
+		}
+		// TODO: detect exit code from threads
+		// c.C() <- &dap.ExitedEvent{
+		// 	Event: dap.Event{
+		// 		Event: "exited",
+		// 	},
+		// 	Body: dap.ExitedEventBody{
+		// 		ExitCode: exitCode,
+		// 	},
+		// }
+	})
+	d.srv.Stop()
+
+	err := d.eg.Wait()
+	d.eg = nil
+	return err
+}
+
+func (d *Adapter) Initialize(c Context, req *dap.InitializeRequest, resp *dap.InitializeResponse) error {
+	close(d.initialized)
+
+	// Set capabilities.
+	resp.Body.SupportsConfigurationDoneRequest = true
+	return nil
+}
+
+type LaunchConfig struct {
```

Comment:
Will this be used for build flags as well? (e.g. the user can write the build flags to launch.json instead of to the global editor config file) If so, the Launch handler will somehow need to be able to trigger the build.

Reply:
> My initial thought was to have the editor translate the JSON launch arguments to the appropriate CLI arguments.

:+1:

> I have to see how the VSCode side works though to see what is better.

IIUC, in all of nvim, VSCode([`startDebugging API`](https://code.visualstudio.com/api/references/vscode-api)), and emacs ([dap-mode](https://github.com/emacs-lsp/dap-mode/blob/b97756665709bea37b9ffe262c5fa9196f1b4577/docs/page/features.md#launchjson-support)), DAP plugin can receive launch.json contents from LaunchRequest.